packages/account/src/BaseSmartAccount.ts

```diff
@@ -0,0 +1,537 @@
+import { JsonRpcProvider, Provider } from '@ethersproject/providers'
+import { BigNumber, BigNumberish, Signer, BytesLike, ethers, Bytes } from 'ethers'
+import { IBaseSmartAccount } from './interfaces/IBaseSmartAccount'
+import { defaultAbiCoder, keccak256, arrayify, Result } from 'ethers/lib/utils'
+import { UserOperation, ChainId } from '@biconomy/core-types'
+import { calcPreVerificationGas, DefaultGasLimits } from './utils/Preverificaiton'
+import { NotPromise, packUserOp } from '@biconomy/common'
+import { IBundler, UserOpResponse } from '@biconomy/bundler'
+import { IPaymaster, PaymasterAndDataResponse } from '@biconomy/paymaster'
+import { EntryPoint_v100, Logger } from '@biconomy/common'
+import { BaseSmartAccountConfig, Overrides, TransactionDetailsForUserOp } from './utils/Types'
+import { GasOverheads } from './utils/Preverificaiton'
+import { EntryPoint, EntryPoint__factory } from '@account-abstraction/contracts'
+import { DEFAULT_ENTRYPOINT_ADDRESS } from './utils/Constants'
+import { RPC_PROVIDER_URLS } from '@biconomy/common'
+
+type UserOperationKey = keyof UserOperation
+
+export abstract class BaseSmartAccount implements IBaseSmartAccount {
+  private isDeployed = false
+  bundler?: IBundler // httpRpcClient
+  paymaster?: IPaymaster // paymasterAPI
+  overheads?: Partial<GasOverheads>
+  entryPointAddress!: string
+  accountAddress?: string
+  owner?: Signer // owner is not mandatory for some account implementations
+  index?: number
+  chainId?: ChainId
+  provider: JsonRpcProvider | Provider // Review
+
+  // entryPoint connected to "zero" address. allowed to make static calls (e.g. to getSenderAddress)
+  private readonly entryPointView!: EntryPoint
+
+  smartAccountConfig: BaseSmartAccountConfig
+
+  constructor(_smartAccountConfig: BaseSmartAccountConfig) {
+    // Review
+    this.smartAccountConfig = _smartAccountConfig
+
+    this.owner = _smartAccountConfig.owner
+    this.index = _smartAccountConfig.index ?? 0
+    this.overheads = _smartAccountConfig.overheads
+    this.entryPointAddress = _smartAccountConfig.entryPointAddress ?? DEFAULT_ENTRYPOINT_ADDRESS
+    this.accountAddress = _smartAccountConfig.accountAddress
```

Comment:
accountAddress should not be passed as constructor argument,  but should be returned by a function getAccountAddress based on counterfactual address.

Reply:
When will someone wants to override the account address? It derived from the owner information and version of BSA.. if someone gives a wrong value then transactions won't work or will start failing in estimations.

Reply:
if someone just wants to make a backend instance with mock singer or no signer. Acme has had such use cases in the past when we had to go back and rework